[WIP] Support for Continuous variables

ankurankan opened the pull request:
- [x] Base classes for continuous variables
- [x] Support for Gaussian variables [WIP]
- [ ] Change Factor and CPD classes to work with continuous variables
- [ ] Inference algorithms working with continuous variables 

### PS: Merge using `--no-ff` option. I would like to have all the changes in a single commit in the history.

ankurankan committed:
Adds GaussianDistribution class:

1. Add a new class for Gaussian Distributions.
2. Adds tests for Gaussian Distribution.

ankurankan committed:
Deletes Canonical Distribution:

There's no need to have two representations of Gaussian Distribution, so deleting Canonical Distribution

ankurankan commented:
@khalibartan @yashu-seth Reviews please. More changes coming, but it will be a lot to review in a single go.

ankurankan committed:
Removes ContinuousFactor class:

1. ContinuousFactor is not supposed to be doing things required for CustomDistribution.
2. CustomDistribution is almost same as ContinuousFactor

khalibartan commented:
@ankurankan My end sem exams are going on, so I don't I will be able to get enough time to review the PR I'll review some parts on Thursday.

ankurankan committed:
adds ContinuousFactor class and tests

ankurankan committed:
DynamicBayesian Networks should accept continuous factor as network parameters

ankurankan committed:
Adds LinearGaussianCPD class

ankurankan committed:
Adds Linear Gaussian CPD

khalibartan commented:
@ankurankan Is there any particular reason for supporting creating distributions with string directly?

khalibartan commented on pgmpy/factors/continuous/LinearGaussianCPD.py:
```diff
@@ -29,106 +31,56 @@ class LinearGaussianCPD(BaseFactor):
     p(Y |x) = N(β0 + β.T * x ; σ2)
```
@ankurankan Can you modify the formula to use μ instead? Feels more relatable with mean.

khalibartan commented on pgmpy/factors/continuous/LinearGaussianCPD.py:
```diff
@@ -29,106 +31,56 @@ class LinearGaussianCPD(BaseFactor):
 
         Examples
         --------
+        For representing the following CPD:
```
@ankurankan I think we should put this in comment. Allows to directly copy paste the doc in ipython console.

khalibartan commented on pgmpy/factors/continuous/LinearGaussianCPD.py:
```diff
@@ -29,106 +31,56 @@ class LinearGaussianCPD(BaseFactor):
-        'Y'
-        >>> copy_cpd.evidence
-        ['X1', 'X2', 'X3']
+        >>> cpd = LinearGaussianCPD(variable='Y', mean=[2, 3, -4, 1], cov=1)
```
@ankurankan Needs both examples with dist and without it.

khalibartan commented on pgmpy/factors/distributions/GaussianDistribution.py:
```diff
@@ -1,581 +1,675 @@
+        self.mean = np.atleast_2d(mean).reshape(self.variables.size, self.evidence.size + 1)
+
+        # Both variables and evidence should be either a str or 1-D array.
+        if not self.variables.ndim == 1:
```
@ankurankan Can you put an empty line before each `elif block`?

khalibartan commented:
@ankurankan I have reviewed changes except for GaussianDistribution and below. It will take me some time for that.There are a lot of changes in Gaussian Distribution. Also, have you tested the Dynamic Bayesian Network with continuous distributions?

ankurankan committed:
Change the structure to make things more uniform.

Now the distributions folders are inside both continuous and discrete. And changes related to that with tests

harishkashyap commented on pgmpy/factors/continuous/distributions/CustomDistribution.py:
```diff
@@ -1,7 +1,7 @@
-from pgmpy.factors.distributions import BaseDistribution
+from pgmpy.factors import BaseDistribution
 
 
```
@ankurankan, minor: shouldn't the dirichlet be x1^alpha-1, x2^beta-1?

harishkashyap replied:
I would think that each distribution, such as dirichlet or beta could be added into the repo itself instead of custom distribution class as they would be useful to all given that they are conjugate priors that are always used. thoughts @ankurankan ?

yashu-seth commented:
@ankurankan The reason I had written a separate class with canonical factor implementation was because they have direct results for sum product operations which would heavily be used in the Inference methods. It was intended to be an internal representation to facilitate Inference on Gaussian networks because conversion of Gaussian distribution to canonical factors and vice versa has direct methods. What do you think?

harishkashyap commented:
@yashu-seth, makes sense. perhaps a different name then?

ankurankan committed:
Changes dir structure of discrete factors to match with continuous factor

ankurankan committed:
Update tests

harishkashyap reviewed (commented):
@ankurankan This is amazing work. I just have a question about discretizing the continuous variables. what is the intention here? We might have to block any hybrid models where discrete variables could be used along with other discrete CPDs as per Daphne's statement "Basic
hybrid network
continuity assumptions that hold in almost all domains imply certain relationships that hold between probabilities associated with “nearby” discretized values of a continuous variable. However, such constraints are very hard to capture in a discrete distribution, where there is no notion that two values of the variable are “close” to each other."
